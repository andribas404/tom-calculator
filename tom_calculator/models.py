"""Models."""
import logging
from typing import Type

import sqlalchemy as sa
from sqlalchemy.dialects.postgresql import UUID

from tom_calculator.database import Base

logger = logging.getLogger(__name__)

TBase = Type[Base]


class Tax(Base):
    """Tax model."""
    __tablename__ = 'taxes'
    __table_args__ = {'comment': 'Table with tax rates per state.'}

    id = sa.Column(
        sa.Integer,
        primary_key=True,
        doc='Tax identifier.',
        comment='Tax identifier.',
    )
    state_name = sa.Column(
        sa.String(20),
        nullable=False,
        index=True,
        unique=True,
        doc='Name of the state.',
        comment='Name of the state.',
    )
    rate = sa.Column(
        sa.Numeric(16, 2),
        nullable=False,
        doc='Rate for the current state measured in `%`.',
        comment='Rate for the current state measured in `%`.',
    )


class Discount(Base):
    """Discount model."""
    __tablename__ = 'discounts'
    __table_args__ = {'comment': 'Table with discount rates applied to the amount.'}

    id = sa.Column(
        sa.Integer,
        primary_key=True,
        doc='Dicsount identifier.',
        comment='Dicsount identifier.',
    )
    amount = sa.Column(
        sa.Numeric(16, 2),
        nullable=False,
        index=True,
        doc='Amount for which this discount is applicable.',
        comment='Amount for which this discount is applicable.',
    )
    rate = sa.Column(
        sa.Numeric(16, 0),
        nullable=False,
        doc='Discount rate applied to amount measured in `%`.',
        comment='Discount rate applied to amount measured in `%`.',
    )


class Order(Base):
    """Order model."""
    __tablename__ = 'orders'
    __table_args__ = {'comment': 'Table with orders.'}
    __mapper_args__ = {'eager_defaults': True}

    id = sa.Column(  # type: ignore[misc]
        UUID,
        primary_key=True,
        server_default=sa.func.uuid_generate_v4(),
        doc='Order identifier.',
        comment='Order identifier.',
    )
    ts = sa.Column(  # type: ignore[misc]
        sa.DateTime,
        server_default=sa.func.now(),
        nullable=False,
        index=True,
        doc='Order timestamp.',
        comment='Order timestamp.',
    )
    amount = sa.Column(
        sa.Numeric(16, 2),
        nullable=False,
        doc='Amount of order received from calculator.',
        comment='Amount of order received from calculator.',
    )
    # after_discount = amount - discount
    after_discount = sa.Column(
        sa.Numeric(16, 2),
        nullable=False,
        doc='Amount of order after discount has been applied.',
        comment='Amount of order after discount has been applied.',
    )
    # tax = after_discount * rate
    tax = sa.Column(
        sa.Numeric(16, 2),
        nullable=False,
        doc='Tax sum calculated from discounted amount.',
        comment='Tax sum calculated from discounted amount.',
    )
    # total = after_discount + tax
    total = sa.Column(
        sa.Numeric(16, 2),
        nullable=False,
        doc='Total amount of order after discount and tax.',
        comment='Total amount of order after discount and tax.',
    )
