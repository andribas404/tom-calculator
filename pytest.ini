[pytest]
addopts =
    --numprocesses auto
    --cov tom_calculator --cov-report term-missing
    --verbose
    --color=yes
    --show-capture=no
    --disable-pytest-warnings

markers =
    unit: Unit tests
    integration: Integration tests
