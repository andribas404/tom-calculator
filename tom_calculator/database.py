import logging
from asyncio import current_task
from contextlib import AbstractAsyncContextManager, asynccontextmanager
from typing import Callable

from sqlalchemy import orm
from sqlalchemy.ext.asyncio import AsyncSession, async_scoped_session, create_async_engine

logger = logging.getLogger(__name__)

Base = orm.declarative_base()

TSession = Callable[..., AbstractAsyncContextManager[orm.Session]]


class Database:
    """Database."""
    def __init__(self, db_dsn: str) -> None:
        self._engine = create_async_engine(db_dsn, future=True, echo=True)
        self._async_session_factory = orm.sessionmaker(
            class_=AsyncSession,
            autocommit=False,
            autoflush=False,
            expire_on_commit=False,
            bind=self._engine,
        )
        self._session = async_scoped_session(self._async_session_factory, scopefunc=current_task)

    async def create_database(self) -> None:
        async with self._engine.begin() as conn:
            await conn.run_sync(Base.metadata.drop_all)
            await conn.run_sync(Base.metadata.create_all)

    async def close(self) -> None:
        await self._engine.dispose()

    @asynccontextmanager
    async def session(self) -> TSession:
        async with self._session() as session:
            try:
                yield session
                await session.commit()
            except Exception:
                logger.exception('Session rollback because of exception')
                await session.rollback()
                raise
            finally:
                await session.close()
