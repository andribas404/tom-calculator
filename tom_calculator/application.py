import logging
from fastapi import FastAPI

from tom_calculator.containers import Container
from tom_calculator.util import get_config_path
from tom_calculator import endpoints

logger = logging.getLogger(__name__)


def create_app() -> FastAPI:
    config_path = get_config_path()
    container = Container()
    container.config.from_yaml(config_path)
    container.wire(modules=[endpoints])

    app = FastAPI()
    app.container = container
    app.include_router(endpoints.router)

    return app


app = create_app()
