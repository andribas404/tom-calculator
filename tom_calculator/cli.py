import asyncio
import os
import subprocess

from dependency_injector.wiring import inject, Provide
import typer

from tom_calculator import services
from tom_calculator.application import create_app

app = typer.Typer()


@inject
def load(
    datadir: str,
    loader_service: services.LoaderService = Provide['loader_service'],
):
    asyncio.run(loader_service.load(datadir))


@app.callback()
def main(
    ctx: typer.Context,
):
    main_app = create_app()
    ctx.obj = main_app


@app.command()
def migrate():
    typer.echo('Starting migration...')
    subprocess.run(['alembic', 'migrate', 'head'])


@app.command()
def migrate_data(
):
    typer.echo('Migrating data...')
    datadir = os.getenv('TOM_DATA')
    load(datadir)


if __name__ == '__main__':
    app()
