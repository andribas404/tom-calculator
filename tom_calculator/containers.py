from dependency_injector import containers, providers

from tom_calculator.database import Database
from tom_calculator.services import OrderService


class Container(containers.DeclarativeContainer):

    config = providers.Configuration()

    db = providers.Singleton(Database, db_dsn=config.db.async_dsn)

    order_service = providers.Factory(
        OrderService,
        session_factory=db.provided.session,
    )
