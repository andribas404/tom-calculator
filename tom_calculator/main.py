"""Main module.

Contains application instance.
"""
import logging

import uvicorn

from tom_calculator.application import create_app

logger = logging.getLogger(__name__)

app = create_app()


if __name__ == '__main__':
    uvicorn.run(app, host='0.0.0.0', port=8000)
