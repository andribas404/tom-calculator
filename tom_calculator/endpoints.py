"""API endpoints."""
import logging
from typing import Any
import uuid

from dependency_injector.wiring import Provide, inject
from fastapi import APIRouter, Depends, HTTPException, Response, status
from starlette.responses import RedirectResponse

from tom_calculator import schemas, services
from tom_calculator.containers import Container

logger = logging.getLogger(__name__)

router = APIRouter()


@router.get('/')
async def root() -> Any:
    """Root endpoint.

    Redirects to openapi.
    """
    response = RedirectResponse(url='/docs#/default/order_create_order_post')
    return response


@router.post(
    '/order',
    status_code=status.HTTP_201_CREATED,
    response_model=schemas.OrderItemOut,
)
@inject
async def order_create(
    item: schemas.CalculatorIn,
    order_service: services.OrderService = Depends(Provide[Container.order_service]),
) -> Any:
    """Endpoint for creating orders from calculator input.

    Returns created order.
    """
    try:
        return await order_service.create(item)
    except services.TaxNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f'Tax for state {item.state_name} not exist.'
        )


@router.get(
    '/order/{item_id}',
    response_model=schemas.OrderItemOut,
)
@inject
async def order_get(
    item_id: uuid.UUID,
    order_service: services.OrderService = Depends(Provide[Container.order_service]),
) -> Any:
    """Endpoint for retrieving order by id."""
    try:
        return await order_service.get(item_id)
    except services.OrderNotFoundError:
        return Response(status_code=status.HTTP_404_NOT_FOUND)
