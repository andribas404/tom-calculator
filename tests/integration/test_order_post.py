"""Test /order endpoint."""
import pytest

from tom_calculator.models import Discount, Tax


class TestOrderPost:
    """Test order creation."""

    @pytest.fixture(autouse=True)
    async def _setup(self, app):
        """Setup entities."""
        async with app.container.db().session() as session:
            tax = Tax(state_name='state', rate=5)
            discount = Discount(amount=100, rate=3)
            session.add(tax)
            session.add(discount)

    @pytest.mark.parametrize(('payload', 'expected'), [
        (
            {"items": [{"quantity": 0, "price": 0}], "state_name": "state"},
            {
                "amount": 0,
                "after_discount": 0,
                "tax": 0,
                "total": 0,
            }
        ),
        (
            {"items": [{"quantity": 0, "price": 0}], "state_name": "state"},
            {
                "amount": 0,
                "after_discount": 0,
                "tax": 0,
                "total": 0,
            }
        ),
        (
            {"items": [{"quantity": 100, "price": 12.34}], "state_name": "state"},
            {
                "amount": 1234.0,
                "after_discount": 1196.98,
                "tax": 59.85,
                "total": 1137.13,
            }
        ),
        (
            {"items": [{"quantity": 7, "price": 0.99}], "state_name": "state"},
            {
                "amount": 6.93,
                "after_discount": 6.93,
                "tax": 0.35,
                "total": 6.58,
            }
        ),
    ])
    def test_valid(self, client, payload, expected):
        """Test valid order creation."""
        response = client.post('/order', json=payload)
        assert response.status_code == 201
        data = response.json()
        data.pop('id')
        data.pop('ts')
        assert data == expected
