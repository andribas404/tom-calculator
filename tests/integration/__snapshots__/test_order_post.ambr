# name: TestOrderPost.test_amount_equal_to_discount[payload0-201]
  <class 'dict'> {
    'after_discount': 100.0,
    'amount': 100,
    'tax': 5.0,
    'total': 95.0,
  }
---
# name: TestOrderPost.test_amount_equal_to_discount[payload1-201]
  <class 'dict'> {
    'after_discount': 700.0,
    'amount': 1000,
    'tax': 35.0,
    'total': 665.0,
  }
---
# name: TestOrderPost.test_amount_greater_than_discount[payload0-201]
  <class 'dict'> {
    'after_discount': 100.01,
    'amount': 100.01,
    'tax': 5.01,
    'total': 95.0,
  }
---
# name: TestOrderPost.test_amount_less_than_discount[payload0-201]
  <class 'dict'> {
    'after_discount': 99.9,
    'amount': 99.9,
    'tax': 5.0,
    'total': 94.9,
  }
---
# name: TestOrderPost.test_best_discount_applied[payload0-201]
  <class 'dict'> {
    'after_discount': 2100.0,
    'amount': 3000,
    'tax': 105.0,
    'total': 1995.0,
  }
---
# name: TestOrderPost.test_empty_list[payload0-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
        ],
        'msg': 'Empty sequence is not allowed.',
        'type': 'assertion_error',
      },
    ],
  }
---
# name: TestOrderPost.test_incorrect_schema[payload0-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
        ],
        'msg': 'field required',
        'type': 'value_error.missing',
      },
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'state_name',
        ],
        'msg': 'field required',
        'type': 'value_error.missing',
      },
    ],
  }
---
# name: TestOrderPost.test_incorrect_schema[payload1-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
        ],
        'msg': 'field required',
        'type': 'value_error.missing',
      },
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'state_name',
        ],
        'msg': 'field required',
        'type': 'value_error.missing',
      },
    ],
  }
---
# name: TestOrderPost.test_incorrect_schema[payload2-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'price',
        ],
        'msg': 'field required',
        'type': 'value_error.missing',
      },
    ],
  }
---
# name: TestOrderPost.test_incorrect_schema[payload3-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          1,
          'quantity',
        ],
        'msg': 'field required',
        'type': 'value_error.missing',
      },
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          1,
          'price',
        ],
        'msg': 'field required',
        'type': 'value_error.missing',
      },
    ],
  }
---
# name: TestOrderPost.test_incorrect_schema[payload4-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'state_name',
        ],
        'msg': 'field required',
        'type': 'value_error.missing',
      },
    ],
  }
---
# name: TestOrderPost.test_negative_price[payload0-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'price',
        ],
        'msg': 'Negative number is not allowed.',
        'type': 'assertion_error',
      },
    ],
  }
---
# name: TestOrderPost.test_negative_quantity[payload0-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'quantity',
        ],
        'msg': 'Negative number is not allowed.',
        'type': 'assertion_error',
      },
    ],
  }
---
# name: TestOrderPost.test_price_correct[payload0-201]
  <class 'dict'> {
    'after_discount': 0.0,
    'amount': 0.0,
    'tax': 0.0,
    'total': 0.0,
  }
---
# name: TestOrderPost.test_price_correct[payload1-201]
  <class 'dict'> {
    'after_discount': 0.0,
    'amount': 0.0,
    'tax': 0.0,
    'total': 0.0,
  }
---
# name: TestOrderPost.test_price_not_float[payload0-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'price',
        ],
        'msg': 'value is not a valid decimal',
        'type': 'type_error.decimal',
      },
    ],
  }
---
# name: TestOrderPost.test_price_not_float[payload1-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'price',
        ],
        'msg': 'none is not an allowed value',
        'type': 'type_error.none.not_allowed',
      },
    ],
  }
---
# name: TestOrderPost.test_price_not_money[payload0-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'price',
        ],
        'msg': 'Not the money format.',
        'type': 'assertion_error',
      },
    ],
  }
---
# name: TestOrderPost.test_price_not_money[payload1-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'price',
        ],
        'msg': 'Not the money format.',
        'type': 'assertion_error',
      },
    ],
  }
---
# name: TestOrderPost.test_price_not_money[payload2-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'price',
        ],
        'msg': 'Not the money format.',
        'type': 'assertion_error',
      },
    ],
  }
---
# name: TestOrderPost.test_price_not_money[payload3-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'price',
        ],
        'msg': 'Not the money format.',
        'type': 'assertion_error',
      },
    ],
  }
---
# name: TestOrderPost.test_price_not_money[payload4-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'price',
        ],
        'msg': 'Not the money format.',
        'type': 'assertion_error',
      },
    ],
  }
---
# name: TestOrderPost.test_price_not_money[payload5-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'price',
        ],
        'msg': 'Not the money format.',
        'type': 'assertion_error',
      },
    ],
  }
---
# name: TestOrderPost.test_quantity_not_int[payload0-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'quantity',
        ],
        'msg': 'value is not a valid integer',
        'type': 'type_error.integer',
      },
    ],
  }
---
# name: TestOrderPost.test_quantity_not_int[payload1-422]
  <class 'dict'> {
    'detail': <class 'list'> [
      <class 'dict'> {
        'loc': <class 'list'> [
          'body',
          'items',
          0,
          'quantity',
        ],
        'msg': 'none is not an allowed value',
        'type': 'type_error.none.not_allowed',
      },
    ],
  }
---
# name: TestOrderPost.test_state_not_exists[payload0-400]
  <class 'dict'> {
    'detail': 'Tax for state state1 not exist.',
  }
---
# name: TestOrderPost.test_state_not_exists[payload1-400]
  <class 'dict'> {
    'detail': 'Tax for state st not exist.',
  }
---
# name: TestOrderPost.test_valid[payload0]
  <class 'dict'> {
    'after_discount': 0.0,
    'amount': 0,
    'tax': 0.0,
    'total': 0.0,
  }
---
# name: TestOrderPost.test_valid[payload1]
  <class 'dict'> {
    'after_discount': 0.0,
    'amount': 0,
    'tax': 0.0,
    'total': 0.0,
  }
---
# name: TestOrderPost.test_valid[payload2]
  <class 'dict'> {
    'after_discount': 863.8,
    'amount': 1234.0,
    'tax': 43.19,
    'total': 820.61,
  }
---
# name: TestOrderPost.test_valid[payload3]
  <class 'dict'> {
    'after_discount': 6.93,
    'amount': 6.93,
    'tax': 0.35,
    'total': 6.58,
  }
---
